### ISSUE TITLE
Wrong calculation or wrong parentheses

### ISSUE THREAD

**Issue by adrianspecker** [davidbkemp/quantum-gate-playground#1]
Hi

many thanks for this impressive illustration - one can learn a lot.

It seems to me and I verified it, that the Y-Operators (and maybe others) - althoug the phase and the circle are illustrated correct - the calculation shown above is not right. So if you produce a complex number c1 for state 0> and a complex c2 for state 1> (e.g.   0.1582+0.5489i |0> - 0.6509+0.5i |1>) and than operate Y it doesn't show the right signs +- (  -0.5+0.6509i |0> - 0.5489-0.1582i |1>)

Best wishes and thanks again!
Adrian

**Comment by davidbkemp** [comment]
Thanks for pointing this out Adrian. I will hopefully find time to look at it this weekend.

**Comment by adrianspecker** [comment]
Thank you for your reply David and thanks again for your visualization! 

I didn't want to make a "change request" and would never ever expect any neartime action on your side - especially not that you now feel under any pressure or even have to work during your weekend because of my post ;-). 

Best wishes from Switzerland, Adrian.

**Comment by davidbkemp** [comment]
My guess is that the problem is to do with how I display the values.
In hindsight I can see that I made a bad choice.
Would you expect `|0> - a+bi |1>` to be equivalent to `|0> - (a+bi)|1>`  or to `|0> - (a-bi)|1>`

I might add paraenthesis when appropriate to make it more clear.
But after my 4 week vacation :-)

**Comment by adrianspecker** [comment]
Hi David

Yes, the second state |1> you could “save” with the introduction of the parenthesis as it would change the sign.
But I fear, that it doesn’t work with the state |0>…

If you create the following state:

Initial State:
0.6066+0.1938i |0> - 0.2169+0.7399i |1>

Than you perform the Operator Y – and it results in:

-0.7399+0.2169i |0> - 0.1938-0.6066i |1>

In my opinion it should show e.g:
0.7399 + 0.2169i  |0> – 0.1938 + 0.6066i |1>

((or could show with paranthesis: (0.7399 + 0.2169i)  |0> – (0.1938 - 0.6066i) |1>))


So probably it is much better to deal without parantheses as one can not “heal” the wrong sign of the first number 0.7399 (in the example) with paranthesis. It shows -0.7399 but should show 0.7399..

So I really believe it is much easier to change the signs without parenthesis. Somehow the calculation of the complex numbers seems to be wrong ? I didn’t check all the other operators but I think there are some similar sign problems. Probably it is just a central subroutine fault.

All the best and relaxing holidays

Adrian







Von: David Kemp [mailto:notifications@github.com]
Gesendet: Samstag, 20. Mai 2017 11:00
An: davidbkemp/quantum-gate-playground <quantum-gate-playground@noreply.github.com>
Cc: Specker Adrian <adrian.specker@fhnw.ch>; State change <state_change@noreply.github.com>
Betreff: Re: [davidbkemp/quantum-gate-playground] Wrong calculation or wrong parentheses (#1)


My guess is that the problem is to do with how I display the values.
In hindsight I can see that I made a bad choice.
Would you expect |0> - a+bi |1> to be equivalent to |0> - (a+bi )|1> or to |0> - (a-bi)|1>

I might add paraenthesis when appropriate to make it more clear.
But after my 4 week vacation :-)

**Comment by davidbkemp** [comment]
@adrianspecker In case you are still curious, I have reworked my quantum computation simulation software.  I believe the root cause of the confusion was that I originally attempting to avoid wrapping the amplitudes in parenthesis.  I made the rather flawed decision to represent

```
- (a + bi)
```

as

```
- a+bi
```

So, now I have gone to the other extreme and always wrap the amplitudes in parenthesis.  This unfortunately leads to ugly scenarios where:

```
0.7071 |0>  - 0.7071 |1>
```

is formatted as:

```
(0.7071)|0> + (-0.7071)|1>
```

I could handle special cases like this, but I will leave it as it is for now until more issues are raised.  :-)

**Comment by adrianspecker** [comment]
Hello David

Many thanks for your efforts, that seems to be a good solution!

Best wishes
Adrian

Von: David Kemp [mailto:notifications@github.com]
Gesendet: Sonntag, 30. Juli 2017 05:41
An: davidbkemp/quantum-gate-playground <quantum-gate-playground@noreply.github.com>
Cc: Specker Adrian <adrian.specker@fhnw.ch>; Mention <mention@noreply.github.com>
Betreff: Re: [davidbkemp/quantum-gate-playground] Wrong calculation or wrong parentheses (#1)


@adrianspecker<https://github.com/adrianspecker> In case you are still curious, I have reworked my quantum computation simulation software. I believe the root cause of the confusion was that I originally attempting to avoid wrapping the amplitudes in parenthesis. I made the rather flawed decision to represent

- (a + bi)

as

- a+bi

So, now I have gone to the other extreme and always wrap the amplitudes in parenthesis. This unfortunately leads to ugly scenarios where:

0.7071 |0>  - 0.7071 |1>

is formatted as:

(0.7071)|0> + (-0.7071)|1>

I could handle special cases like this, but I will leave it as it is for now until more issues are raised. :-)
